### PR TITLE
Remove fprint methods from ast.numpyext

### PR DESCRIPTION
**Pull request by mostafaMamoni**
* Remove fprint methods from `ast.numpyext `and `ast.core` and reimplement them in `Fcodeprinter`;
* Add a new AST node `NumpyAutoFill`, with `NumpyZeros` and `NumpyOnes` inheriting from it instead of `NumpyEmpty`, in order to:
   - Correctly print `NumpyEmpty` objects in an `Assign` node without catching instances of `NumpyZeros` and `NumpyOnes`;
   - Avoid repeating the same __new__ function as `NumpyEmpty`.

### CODE REVIEW

File: pyccel/codegen/printing/fcode.py [hunk 1]
```diff
@@ -1089,12 +1291,53 @@ def _print_Assign(self, expr):
...
+        
+        if isinstance(rhs, (NumpyComplex, NumpyReal, NumpyArray, NumpyLinspace)):
+            rhs = self._print(rhs)
+            return '{0} = {1}\n'.format(lhs_code,rhs)
+        
+        if isinstance(rhs, (NumpyCross, NumpyDiag, NumpyWhere)):
+            return self._print(rhs, expr.lhs)
+        
+        if isinstance(rhs, PyccelArraySize):
             return rhs.fprint(self._print, expr.lhs) + '\n'
```
Comment (EmilyBourne) [inline pyccel/codegen/printing/fcode.py:1310]: Shouldn't this fprint be handled too?

File: pyccel/codegen/printing/fcode.py [hunk 2]
```diff
@@ -1089,12 +1291,53 @@ def _print_Assign(self, expr):
...
+        if isinstance(rhs, NumpyRand):
+
+            stack_array = False
+            if self._current_function:
+                name = self._current_function
+                func = self.get_function(name)
+                lhs_name = expr.lhs.name
+                vars_dict = {i.name: i for i in func.local_vars}
+                if lhs_name in vars_dict:
+                    stack_array = vars_dict[lhs_name].is_stack_array
```
Comment (EmilyBourne) [inline pyccel/codegen/printing/fcode.py:1322]: This doesn't look like it should be in the printing. Was it there before? @saidctb do you know what this does?

File: pyccel/codegen/printing/fcode.py [hunk 3]
```diff
@@ -1089,12 +1291,53 @@ def _print_Assign(self, expr):
...
+        if isinstance(rhs, NumpyFull):
+
+            stack_array = False
+            if self._current_function:
+                name = self._current_function
+                func = self.get_function(name)
+                lhs_name = expr.lhs.name
+                vars_dict = {i.name: i for i in func.local_vars}
+                if lhs_name in vars_dict:
+                    stack_array = vars_dict[lhs_name].is_stack_array
```
Comment (EmilyBourne) [inline pyccel/codegen/printing/fcode.py:1335]: This code looks identical to the code above. Why not create a function?

**Review by EmilyBourne — CHANGES_REQUESTED** [review]

File: pyccel/codegen/printing/fcode.py [hunk 4]
```diff
@@ -611,26 +611,199 @@ def _print_PythonImag(self, expr):
 
     #========================== Numpy Elements ===============================#
 
-    def _print_NumpySum(self, expr):
-        return expr.fprint(self._print)
+    def _print_NumpySum(self, expr, lhs=None):
```
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/codegen/printing/fcode.py:614]: This is not the only place where this is done

File: pyccel/codegen/printing/fcode.py [hunk 5]
```diff
@@ -611,26 +611,199 @@ def _print_PythonImag(self, expr):
...
+        """Fortran print."""
+
+        rhs_code = self._print(expr.arg)
+        if lhs:
+            lhs_code = self._print(lhs)
+            return '{0} = sum({1})'.format(lhs_code, rhs_code)
```
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/codegen/printing/fcode.py:620]: This is not the only place where this is done

File: pyccel/codegen/printing/fcode.py [hunk 6]
```diff
@@ -1089,12 +1291,53 @@ def _print_Assign(self, expr):
...
+            stack_array = False
+            if self._current_function:
+                name = self._current_function
+                func = self.get_function(name)
+                lhs_name = expr.lhs.name
+                vars_dict = {i.name: i for i in func.local_vars}
+                if lhs_name in vars_dict:
+                    stack_array = vars_dict[lhs_name].is_stack_array
+
+            return self._print_NumpyFull(rhs, lhs_code)
```
Comment (EmilyBourne) [inline pyccel/codegen/printing/fcode.py:1337]: We should never call the `_print_ClassName` functions by name. The calls should always go through a call to `_print`

File: pyccel/codegen/printing/fcode.py [hunk 7]
```diff
@@ -611,26 +611,199 @@ def _print_PythonImag(self, expr):
...
+        order = expr.order
+
+        if lhs is not None:
+            lhs  = self._print(lhs)
+
+            if rank == 2:
+                alloc = 'allocate({0}(0:size({1},1)-1,0:size({1},2)-1))'.format(lhs, first)
+
+            elif rank == 1:
+                alloc = 'allocate({}(0:size({})-1)'.format(lhs, first)
```
[the remark this answers is not in the document]
Comment (mostafaMamoni) [reply pyccel/codegen/printing/fcode.py]: i removed this and the return statement for codacy.

File: pyccel/codegen/printing/fcode.py [hunk 8]
```diff
@@ -705,7 +878,29 @@ def _print_NumpyRandint(self, expr):
...
-        return expr.fprint(self._print)
+        if expr.high is None:
+            randreal = self._print(PyccelMul(expr.low, NumpyRand()))
+        else:
+            randreal = self._print(PyccelAdd(PyccelMul(PyccelAssociativeParenthesis(PyccelMinus(expr.high, expr.low)), NumpyRand()), expr.low))
+
+        prec_code = self._print(expr.precision)
+        return 'floor({}, kind={})'.format(randreal, prec_code)
+    
+    def _print_NumpyFull(self, expr, lhs_code):
```
[the remark this answers is not in the document]
Comment (mostafaMamoni) [reply pyccel/codegen/printing/fcode.py:880]: what about the if statements here?

File: pyccel/codegen/printing/fcode.py (hunk 1 above)
Comment (mostafaMamoni) [reply pyccel/codegen/printing/fcode.py:1301]: its not in ast.numpyext

File: pyccel/codegen/printing/fcode.py [hunk 9]
```diff
@@ -1089,12 +1291,53 @@ def _print_Assign(self, expr):
...
+        if isinstance(rhs, PythonInt):
+            rhs = expr.rhs.fprint(self._print)
+            return '{0} = {1}\n'.format(lhs_code,rhs)
+        
+        if isinstance(rhs, (NumpyComplex, NumpyReal, NumpyArray, NumpyLinspace)):
+            rhs = self._print(rhs)
+            return '{0} = {1}\n'.format(lhs_code,rhs)
+        
+        if isinstance(rhs, (NumpyCross, NumpyDiag, NumpyWhere)):
+            return self._print(rhs, expr.lhs)
```
[the remark this answers is not in the document]
Comment (mostafaMamoni) [reply pyccel/codegen/printing/fcode.py]: you're right for NumpyCross, but the others I need a confirmation from you because I will move many outside the method because of lhs argument.

File: pyccel/codegen/printing/fcode.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (saidctb) [reply pyccel/codegen/printing/fcode.py]: we can delet numpy cross for the moment

File: pyccel/codegen/printing/fcode.py [hunk 10]
```diff
@@ -611,26 +611,199 @@ def _print_PythonImag(self, expr):
...
+
+    def _print_NumpyWhere(self, expr, lhs):
+
+        ind   = self._print(expr.index)
+        mask  = self._print(expr.mask)
+        lhs   = self._print(lhs)
+
+        stmt  = 'pack([({ind},{ind}=0,size({mask})-1)],{mask})'.format(ind=ind,mask=mask)
+        stmt  = '{lhs}(:,0) = {stmt}'.format(lhs=lhs, stmt=stmt)
+        alloc = 'allocate({}(0:count({})-1,0:0))'.format(lhs, mask)
```
[the remark this answers is not in the document]
Comment (saidctb) [reply pyccel/codegen/printing/fcode.py:771]: we can delete them for the moment

File: pyccel/codegen/printing/fcode.py (hunk 2 above)
Comment (saidctb) [reply pyccel/codegen/printing/fcode.py:1307]: this should be removed, it handles stack arrays , so that we don't allocate them

Comment (saidctb) [reply pyccel/codegen/printing/fcode.py:1307]: it's probably not needed anymore

File: pyccel/codegen/printing/fcode.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/codegen/printing/fcode.py:877]: The `if` statement is included to handle the case of `NumpyEmpty`. I think there should be an `if` statement in the `Assign` to handle this case.

File: pyccel/codegen/printing/fcode.py [hunk 11]
```diff
@@ -612,25 +612,186 @@ def _print_PythonImag(self, expr):
...
+        """Fortran print."""
 
-    def _print_NumpyProduct(self, expr):
-        return expr.fprint(self._print)
+        rhs_code = self._print(expr.arg)
+        return 'sum({0})'.format(rhs_code)
 
-    def _print_NumpyMatmul(self, expr):
-        return expr.fprint(self._print)
+    def _print_NumpyProduct(self, expr, lhs=None):
```
Comment (EmilyBourne) [inline pyccel/codegen/printing/fcode.py:620]: See above. `lhs` shouldn't be here

File: pyccel/codegen/printing/fcode.py [hunk 12]
```diff
@@ -612,25 +612,186 @@ def _print_PythonImag(self, expr):
...
+        """Fortran print."""
+
+        rhs_code = self._print(expr.arg)
+        if lhs:
+            lhs_code = self._print(lhs)
+            return '{0} = product({1})'.format(lhs_code, rhs_code)
+        return 'product({0})'.format(rhs_code)
+
+    def _print_NumpyMatmul(self, expr, lhs=None):
```
Comment (EmilyBourne) [inline pyccel/codegen/printing/fcode.py:629]: See above. `lhs` shouldn't be here

File: pyccel/codegen/printing/fcode.py [hunk 13]
```diff
@@ -612,25 +612,186 @@ def _print_PythonImag(self, expr):
...
+                         a[2]*b[0]-a[0]*b[2],
+                         a[0]*b[1]-a[1]*b[0]]
+
+        cross_product = PythonTuple(cross_product)
+        cross_product = self._print(cross_product)
+        first = self._print(expr.first)
+        order = expr.order
+
+        if lhs is not None:
+            lhs  = self._print(lhs)
```
Comment (EmilyBourne) [inline pyccel/codegen/printing/fcode.py:695]: `lhs` is no longer an argument, so you can't use it in your code

File: pyccel/codegen/printing/fcode.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/codegen/printing/fcode.py:771]: I suppose we can always collect the implementation from the history when we want to implement it. So then the following classes should be removed:
- NumpyCross
- NumpyWhere
- NumpyDiag
? @mostafaMamoni have I missed any classes which need the `lhs`?

File: pyccel/codegen/printing/fcode.py [hunk 14]
```diff
@@ -705,7 +866,28 @@ def _print_NumpyRandint(self, expr):
         if expr.rank != 0:
             errors.report(FORTRAN_ALLOCATABLE_IN_EXPRESSION,
                           symbol=expr, severity='fatal')
-        return expr.fprint(self._print)
+        if expr.high is None:
+            randreal = self._print(PyccelMul(expr.low, NumpyRand()))
+        else:
+            randreal = self._print(PyccelAdd(PyccelMul(PyccelAssociativeParenthesis(PyccelMinus(expr.high, expr.low)), NumpyRand()), expr.low))
```
Comment (EmilyBourne) [inline pyccel/codegen/printing/fcode.py:872]: Is `PyccelAssociativeParenthesis` still needed here? It should be added automatically now

File: pyccel/codegen/printing/fcode.py [hunk 15]
```diff
@@ -705,7 +866,28 @@ def _print_NumpyRandint(self, expr):
...
+
+        prec_code = self._print(expr.precision)
+        return 'floor({}, kind={})'.format(randreal, prec_code)
```
Comment (EmilyBourne) [inline pyccel/codegen/printing/fcode.py:875]: Does this print the correct ISO_C precision?

File: pyccel/codegen/printing/fcode.py (hunk 1 above)
Comment (EmilyBourne) [reply pyccel/codegen/printing/fcode.py:1295]: ok

File: pyccel/codegen/printing/fcode.py (hunk 2 above)
Comment (EmilyBourne) [reply pyccel/codegen/printing/fcode.py:1307]: Given that allocation is handled with the `Allocate` node that sounds right to me. @mostafaMamoni this code, and the identical code below it should be removed

File: pyccel/codegen/printing/fcode.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/codegen/printing/fcode.py]: So I have checked this locally. With a few small fixes, the `NumpyCross` class still works. However it is not tested at all, only a subset of `numpy.cross` is actually implemented, and I think that in the future the function will be implemented using the method in PR #562 . Given that fact, I agree that we should just remove the `NumpyCross` class for now

File: pyccel/codegen/printing/fcode.py [hunk 16]
```diff
@@ -705,7 +853,21 @@ def _print_NumpyRandint(self, expr):
...
+        prec_code = self._print(iso_c_binding["integer"][expr.precision])
+        return 'floor({}, kind={})'.format(randreal, prec_code)
+
+    def _print_NumpyFull(self, expr):
+
+        # Create statement for initialization
+        if expr.fill_value is not None:
+            init_value = self._print(expr.fill_value)
+
+        return init_value
```
Comment (EmilyBourne) [inline pyccel/codegen/printing/fcode.py:870]: If `expr.fill_value is None` then `init_value` has not been initialised and this function call will crash

File: pyccel/codegen/printing/fcode.py [hunk 17]
```diff
@@ -1089,23 +1256,40 @@ def _print_Assign(self, expr):
...
-                    stack_array = vars_dict[lhs_name].is_stack_array
+        if isinstance(rhs, NumpyFull):
+            stmts = []
+            if rhs.fill_value is not None:
+                code_init = '{0} = {1}'.format(lhs_code, self._print(rhs))
+                stmts.append(code_init)
+            if len(stmts) == 0:
+                return ''
+            else:
+                return '\n'.join(stmts) + '\n'
```
Comment (EmilyBourne) [inline pyccel/codegen/printing/fcode.py:1288]: This can be much simpler:
```
if isinstance(rhs, NumpyEmpty):
    return ''
```

File: pyccel/codegen/printing/fcode.py [hunk 18]
```diff
@@ -1089,23 +1256,40 @@ def _print_Assign(self, expr):
...
 
-            return rhs.fprint(self._print, expr.lhs, stack_array)
+        if isinstance(rhs, (NumpyFullLike, NumpyEmptyLike,\
+						NumpyZerosLike, NumpyOnesLike)):
+            return rhs.fprint(self._print, expr.lhs)
```
Comment (EmilyBourne) [inline pyccel/codegen/printing/fcode.py:1292]: Why does this `fprint` still exist?

File: pyccel/codegen/printing/fcode.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (mostafaMamoni) [reply pyccel/codegen/printing/fcode.py]: No, for NumpyFullLike, NumpyOnesLike, NumpyZeroesLike and NumpyEmpyLlike will be handled as well as NumpyFull.

**Comment by EmilyBourne** [comment]
The test `tests/errors/codegen/fortran/FORTRAN_ALLOCATABLE_IN_EXPRESSION.py` is failing. This was supposed to raise an error about allocatables in expressions, however it is no longer needed now that we are using an `Allocatable` node. The code:
```
import numpy as np
a = np.ones(4)*5
```
is correctly translated to:
```
program prog_FORTRAN_ALLOCATABLE_IN_EXPRESSION

use ISO_C_BINDING

implicit none

real(C_DOUBLE), allocatable :: a(:)

allocate(a(0:4_C_INT64_T - 1_C_INT64_T))
a = 1.0_C_DOUBLE * 5_C_INT64_T

end program prog_FORTRAN_ALLOCATABLE_IN_EXPRESSION
```
This test can therefore be removed.

File: pyccel/codegen/printing/fcode.py (hunk 18 above)
Comment (mostafaMamoni) [reply pyccel/codegen/printing/fcode.py:1292]: because their fprints don't exit in ast.numpyext, Is the printer of NumpyFull will be enough for those? if so they will be handled with _print_NumpyFull with some modifications.

File: pyccel/codegen/printing/fcode.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/codegen/printing/fcode.py]: You will need to make one additional modification to deal with stupid users ;) . If a user does:
```
a = np.empty(3)*5
```
Then [this error](https://github.com/pyccel/pyccel/blob/master/pyccel/codegen/printing/fcode.py#L2574-L2575) should be raised. This can be done by adding a specialised `_print_NumpyEmpty` function now that `NumpyOnes` and `NumpyZeros` no longer inherit from `NumpyEmpty`. You could modify `tests/errors/codegen/fortran/FORTRAN_ALLOCATABLE_IN_EXPRESSION.py` to test this

File: pyccel/codegen/printing/fcode.py [hunk 19]
```diff
@@ -612,25 +612,173 @@ def _print_PythonImag(self, expr):
...
+
+        # C ordering
+        return 'matmul({1},{0})'.format(a_code, b_code)
+
+    def _print_NumpyImag(self, expr):
+        """Fortran print."""
+
+        value = self._print(expr.arg)
+        code = 'aimag({0})'.format(value)
+        return code
```
Comment (EmilyBourne) [inline pyccel/codegen/printing/fcode.py:647]: This class inherits from `PythonImag` now. Does it need a specialised `_print_NumpyImag` method or does the `_print_PythonImag` function do the same thing?

File: pyccel/codegen/printing/fcode.py (hunk 16 above)
Comment (EmilyBourne) [reply pyccel/codegen/printing/fcode.py:870]: If we never expect `expr.fill_value` to be `None` then the `if` is unnecessary

File: pyccel/codegen/printing/fcode.py (hunk 1 above)
Comment (EmilyBourne) [reply pyccel/codegen/printing/fcode.py:1275]: Do you know if there are many `fprint` functions outside of `ast.numpyext`? If there aren't very many then it might not be worth doing this in 2 PRs

File: pyccel/codegen/printing/fcode.py [hunk 20]
```diff
@@ -1089,23 +1255,23 @@ def _print_Assign(self, expr):
...
-            if self._current_function:
-                name = self._current_function
-                func = self.get_function(name)
-                lhs_name = expr.lhs.name
-                vars_dict = {i.name: i for i in func.local_vars}
-                if lhs_name in vars_dict:
-                    stack_array = vars_dict[lhs_name].is_stack_array
+        if isinstance(rhs, (NumpyWhere, NumpyCross, NumpyComplex, NumpyReal, NumpyArray, NumpyLinspace)):
+            rhs = self._print(rhs)
+            return '{0} = {1}\n'.format(lhs_code,rhs)
```
Comment (EmilyBourne) [inline pyccel/codegen/printing/fcode.py:1268]: This block isn't needed. The same calls occur if you remove this code

File: pyccel/codegen/printing/fcode.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/codegen/printing/fcode.py]: Ok. In fact it seems that `NumpyWhere` also doesn't need the `lhs` so there is no need to remove it. So the only classes that need removing are:
- NumpyCross
- NumpyDiag

The whole class can be deleted, not just the implementation of its printing

File: pyccel/codegen/printing/fcode.py [hunk 21]
```diff
@@ -1089,12 +1291,53 @@ def _print_Assign(self, expr):
         if isinstance(rhs, (PythonRange, Product)):
             return ''
 
-        if isinstance(rhs, (NumpyArray, NumpyLinspace, NumpyDiag, NumpyCross,\
-						NumpyWhere, PyccelArraySize)):
+        if isinstance(rhs, (PythonLen, NumpyRandint)):
+            rhs_code = self._print(rhs)
+            return '{0} = {1}\n'.format(lhs_code, rhs_code)
```
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/codegen/printing/fcode.py:1260]: We should be able to use the code here: https://github.com/pyccel/pyccel/blob/master/pyccel/codegen/printing/fcode.py#L1169-L1177

File: pyccel/codegen/printing/fcode.py (hunk 20 above)
Comment (EmilyBourne) [reply pyccel/codegen/printing/fcode.py:1268]: See:
https://github.com/pyccel/pyccel/blob/master/pyccel/codegen/printing/fcode.py#L1169-L1177

File: pyccel/codegen/printing/fcode.py (hunk 19 above)
Comment (mostafaMamoni) [reply pyccel/codegen/printing/fcode.py:645]: yeah it does the same as `_print_PythonImag`, so can we remove it?.

Comment (EmilyBourne) [reply pyccel/codegen/printing/fcode.py:645]: When looking for the function to print a `NumpyImag` the `FCodePrinter` will call `_print_NumpyImag` if it exists, and if it doesn't then it will call `_print_PythonImag`

Comment (mostafaMamoni) [reply pyccel/codegen/printing/fcode.py:645]: I think i will keep the code readable and make `_print_NumpyImag` call `_print_PythonImag`, what do you think ?.

Comment (EmilyBourne) [reply pyccel/codegen/printing/fcode.py:645]: I think this is a bad idea. It goes against the pattern established elsewhere. It leads to redundant code, and if we used your method and wanted to rename a class, this would add a lot more changes. For every other class we only implement the functions that are absolutely necessary in order to reduce the amount of code

File: pyccel/codegen/printing/fcode.py [hunk 22]
```diff
@@ -612,25 +610,161 @@ def _print_PythonImag(self, expr):
...
+        elif rank == 1:
+
+            lhs   = IndexedVariable(lhs)[expr.index, expr.index]
+            rhs   = IndexedVariable(expr.array)[expr.index]
+            body  = [Assign(lhs, rhs)]
+            body  = For(expr.index, PythonRange(PythonLen(expr.array)), body)
+            code  = self._print(body)
+            alloc = 'allocate({0}(0: size({1},1)-1, 0: size({1},1)-1))'.format(lhs, array)
+
+        return alloc + '\n' + code
```
Comment (EmilyBourne) [inline pyccel/codegen/printing/fcode.py:712]: Didn't you delete this class?

File: pyccel/codegen/printing/fcode.py [hunk 23]
```diff
@@ -612,25 +610,161 @@ def _print_PythonImag(self, expr):
...
+
+                code = 'reshape({}, shape({}), order=[2, 1])'.format(cross_product, first)
+            else:
+
+                code = 'reshape({}, shape({})'.format(cross_product, first)
+
+        elif rank == 1:
+            code = cross_product
+
+        return code
```
Comment (EmilyBourne) [inline pyccel/codegen/printing/fcode.py:688]: Didn't you delete this class?

File: pyccel/codegen/printing/fcode.py [hunk 24]
```diff
@@ -1060,6 +1206,11 @@ def _print_SymbolicAssign(self, expr):
         stmt = Comment(str(expr))
         return self._print_Comment(stmt)
 
+    def _print_NumpyReal(self, expr):
+        value = self._print(expr.arg)
+        prec = self._print(expr.precision)
```
Comment (EmilyBourne) [inline pyccel/codegen/printing/fcode.py:1211]: Don't we need ISO_C_BINDING precision?

**Review by EmilyBourne — CHANGES_REQUESTED** [review]

File: pyccel/codegen/printing/fcode.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (mostafaMamoni) [reply pyccel/codegen/printing/fcode.py]: should we still need a specialized `_print_NumpyEmpty`after we're working with:
`if isinstance(rhs, NumpyEmpty):
            return ''`
for `tests/errors/codegen/fortran/FORTRAN_ALLOCATABLE_IN_EXPRESSION.py` I removed it ;).

[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/codegen/printing/fcode.py]: Probably we do need a specialized `_print_NumpyEmpty` to make sure the error is raised. If you do:
```
a = np.empty(3)*5
```
Then you have:
```
PyccelAssign(Variable('a'), PyccelMul(NumpyEmpty(LiteralInteger(3)), LiteralInteger(5))
```
So you won't reach the if block `if isinstance(rhs, NumpyEmpty): return ''`

[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/codegen/printing/fcode.py]: > for `tests/errors/codegen/fortran/FORTRAN_ALLOCATABLE_IN_EXPRESSION.py` I removed it ;).

You can get it back by doing:
```
git checkout master -- tests/errors/codegen/fortran/FORTRAN_ALLOCATABLE_IN_EXPRESSION.py
```

File: pyccel/codegen/printing/fcode.py (hunk 22 above)
Comment (mostafaMamoni) [reply pyccel/codegen/printing/fcode.py:712]: this is not a class, but yes I removed `NumpyDiag` in ast.numpyast, and sorry for `lhs` argument I think its cursed for me.

File: pyccel/codegen/printing/fcode.py (hunk 23 above)
Comment (mostafaMamoni) [reply pyccel/codegen/printing/fcode.py:688]: the same with numpydiag

File: pyccel/codegen/printing/fcode.py (hunk 24 above)
Comment (mostafaMamoni) [reply pyccel/codegen/printing/fcode.py:1211]: yes,

File: pyccel/codegen/printing/fcode.py (hunk 22 above)
Comment (EmilyBourne) [reply pyccel/codegen/printing/fcode.py:712]: You are right, it is not a class, but it is a function that is only called on instances of the class `NumpyDiag` so if `NumpyDiag` doesn't exist anymore then this function doesn't need to either

Comment (EmilyBourne) [reply pyccel/codegen/printing/fcode.py:712]: The `lhs` being here is the reason why we said to delete this class. There is no way round it. It's not your fault ^^

File: pyccel/codegen/printing/fcode.py (hunk 1 above)
Comment (mostafaMamoni) [reply pyccel/codegen/printing/fcode.py]: there some of them in builtins for python classes and their calls in fcode.

Comment (EmilyBourne) [reply pyccel/codegen/printing/fcode.py]: With grep I see 4 methods. So I think it can be done in this PR

File: pyccel/codegen/printing/fcode.py [hunk 25]
```diff
@@ -604,47 +602,128 @@ def _print_PythonSum(self, expr):
...
+        return rhs_code
 
     def _print_NumpyFloor(self, expr):
         result_code = self._print_MathFloor(expr)
         return 'real({}, {})'.format(result_code, iso_c_binding["real"][8])
 
     # ======================================================================= #
     def _print_PyccelArraySize(self, expr):
-        return expr.fprint(self._print)
+        init_value = self._print(expr.arg)
```
Comment (EmilyBourne) [inline pyccel/codegen/printing/fcode.py:705]: Did you delete the old fprint method?

File: pyccel/codegen/printing/fcode.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/codegen/printing/fcode.py:1189]: This comment is hidden but still valid

**Review by EmilyBourne — CHANGES_REQUESTED** [review]

File: pyccel/codegen/printing/fcode.py (hunk 25 above)
Comment (mostafaMamoni) [reply pyccel/codegen/printing/fcode.py:705]: yes, you can check the new one in fcode.

Comment (EmilyBourne) [reply pyccel/codegen/printing/fcode.py:705]: Yes I can see the new method. That is where I left the comment. But the old method (`PyccelArraySize.fprint`) is no longer needed. It is defined in ast/core.py but according to github that file hasn't been modified. Are you sure you pushed that change?

Comment (mostafaMamoni) [reply pyccel/codegen/printing/fcode.py:705]: oh, you're right i forgot a fprint in ast.core, it should be removed now.

**Review by EmilyBourne — APPROVED** [review]

File: pyccel/ast/numpyext.py [hunk 26]
```diff
@@ -878,27 +562,10 @@ def fill_value(self):
...
-            return ''
-        else:
-            return '\n'.join(stmts) + '\n'
-
 #==============================================================================
-class NumpyEmpty(NumpyFull):
-    """ Represents a call to numpy.empty for code generation.
+class NumpyAutoFill(NumpyFull):
+    """ Abstract class for all classes which inherit from NumpyFull but
+        the fill_value is implicitly specified
```
Comment (yguclu) [inline pyccel/ast/numpyext.py:568]: Why is this class needed?

Comment (mostafaMamoni) [reply pyccel/ast/numpyext.py:568]: NumpyZeros, NumpyOnes was inheriting from NumpyEmpty, so in the printing NumpyEmpty could be ether NumpyZeros or NumpyOnes in every isinstance block and this not good the new implementation of printing, this class fix that conflict and prevent repeating the same `__new__` function in NumpyEmpty.

Comment (EmilyBourne) [reply pyccel/ast/numpyext.py:568]: >so in the printing NumpyEmpty could be ether NumpyZeros or NumpyOnes in every isinstance block and this not good the new

Put another way, `isinstance` couldn't be used to handle the specific problem of `NumpyEmpty` in an `Assign` as it was also catching the subclasses `NumpyZeros` and `NumpyOnes`

Comment (yguclu) [reply pyccel/ast/numpyext.py:568]: Thanks for the explanation. This should be added to the PR's description.

**Review by yguclu — CHANGES_REQUESTED** [review]
Please add to the PR's description the fact that you created a new AST node class, and the reason for doing so.

File: tests/errors/codegen/fortran/FORTRAN_ALLOCATABLE_IN_EXPRESSION.py [hunk 27]
```diff
@@ -1,3 +1,3 @@
 # pylint: disable=missing-function-docstring, missing-module-docstring/
 import numpy as np
-a = np.ones(4)*5
+a = np.empty(4)*5
```
Comment (saidctb) [inline tests/errors/codegen/fortran/FORTRAN_ALLOCATABLE_IN_EXPRESSION.py:3]: why did you change this example ?

**Review by saidctb — APPROVED** [review]

File: tests/errors/codegen/fortran/FORTRAN_ALLOCATABLE_IN_EXPRESSION.py (hunk 27 above)
Comment (EmilyBourne) [reply tests/errors/codegen/fortran/FORTRAN_ALLOCATABLE_IN_EXPRESSION.py:3]: Now that the allocation has been separated from the printing `ones` is printed as `1` in fortran. So:
```
a = np.ones(4)*5
```
is correctly translated to:
```
allocate(a(0:4))
a = 1*5
```

Comment (saidctb) [reply tests/errors/codegen/fortran/FORTRAN_ALLOCATABLE_IN_EXPRESSION.py:3]: ah okay , perfect

Comment (EmilyBourne) [reply tests/errors/codegen/fortran/FORTRAN_ALLOCATABLE_IN_EXPRESSION.py:3]: As all the values of the allocatable in the expression are the same we don't need a temporary array so there is no problem

**Review by yguclu — APPROVED** [review]
